isisd/isis_sr.c

```diff
@@ -0,0 +1,1654 @@
+		ladj_sid->family = AF_INET6;
+		ladj_sid->flags = EXT_SUBTLV_LINK_ADJ_SID_VFLG
+				  | EXT_SUBTLV_LINK_ADJ_SID_LFLG
+				  | EXT_SUBTLV_LINK_ADJ_SID_FFLG;
+		ladj_sid->weight = 0;
+		memcpy(ladj_sid->neighbor_id, adj->sysid,
+		       sizeof(ladj_sid->neighbor_id));
+		ladj_sid->sid = local_label;
+
+		isis_tlvs_add_lan_adj_sid(circuit->ext, ladj_sid);
+		adj->sr.ipv6.ladj_sid = ladj_sid;
+		break;
+	case CIRCUIT_T_P2P:
+		adj_sid = XCALLOC(MTYPE_ISIS_SUBTLV, sizeof(*adj_sid));
+		adj_sid->family = AF_INET6;
+		adj_sid->flags = EXT_SUBTLV_LINK_ADJ_SID_VFLG
+				 | EXT_SUBTLV_LINK_ADJ_SID_LFLG
+				 | EXT_SUBTLV_LINK_ADJ_SID_FFLG;
+		adj_sid->weight = 0;
+		adj_sid->sid = local_label;
+
+		isis_tlvs_add_adj_sid(circuit->ext, adj_sid);
+		adj->sr.ipv6.adj_sid = adj_sid;
+		break;
+	default:
+		flog_err(EC_LIB_DEVELOPMENT, "%s: unexpected circuit type: %u",
+			 __func__, circuit->circ_type);
+		exit(1);
+	}
+
+	isis_sr_adj_sid_install_uninstall(true, circuit, adj, local_label,
+					  AF_INET6, &adj->ipv6_addresses[0]);
+}
+
+static void isis_sr_circuit_del_ipv4_adj_sid(struct isis_circuit *circuit,
+					     struct isis_adjacency *adj)
+{
+	struct isis_adj_sid *adj_sid;
+	struct isis_lan_adj_sid *ladj_sid;
+
+	switch (circuit->circ_type) {
+	case CIRCUIT_T_BROADCAST:
+		ladj_sid = adj->sr.ipv4.ladj_sid;
+		if (ladj_sid) {
+			isis_sr_adj_sid_install_uninstall(
+				false, circuit, adj, ladj_sid->sid, AF_INET,
+				&adj->ipv4_addresses[0]);
+			isis_tlvs_del_lan_adj_sid(circuit->ext, ladj_sid);
+			adj->sr.ipv4.ladj_sid = NULL;
+		}
+		break;
+	case CIRCUIT_T_P2P:
+		adj_sid = adj->sr.ipv4.adj_sid;
+		if (adj_sid) {
+			isis_sr_adj_sid_install_uninstall(
+				false, circuit, adj, adj_sid->sid, AF_INET,
+				&adj->ipv4_addresses[0]);
+			isis_tlvs_del_adj_sid(circuit->ext, adj_sid);
+			adj->sr.ipv4.adj_sid = NULL;
+		}
+		break;
+	default:
+		flog_err(EC_LIB_DEVELOPMENT, "%s: unexpected circuit type: %u",
+			 __func__, circuit->circ_type);
+		exit(1);
+	}
+}
+
+static void isis_sr_circuit_del_ipv6_adj_sid(struct isis_circuit *circuit,
+					     struct isis_adjacency *adj)
+{
+	struct isis_adj_sid *adj_sid;
+	struct isis_lan_adj_sid *ladj_sid;
+
+	switch (circuit->circ_type) {
+	case CIRCUIT_T_BROADCAST:
+		ladj_sid = adj->sr.ipv6.ladj_sid;
+		if (ladj_sid) {
+			isis_sr_adj_sid_install_uninstall(
+				false, circuit, adj, ladj_sid->sid, AF_INET6,
+				&adj->ipv6_addresses[0]);
+			isis_tlvs_del_lan_adj_sid(circuit->ext, ladj_sid);
+			adj->sr.ipv6.ladj_sid = NULL;
+		}
+		break;
+	case CIRCUIT_T_P2P:
+		adj_sid = adj->sr.ipv6.adj_sid;
+		if (adj_sid) {
+			isis_sr_adj_sid_install_uninstall(
+				false, circuit, adj, adj_sid->sid, AF_INET6,
+				&adj->ipv6_addresses[0]);
+			isis_tlvs_del_adj_sid(circuit->ext, adj_sid);
+			adj->sr.ipv6.adj_sid = NULL;
+		}
+		break;
+	default:
+		flog_err(EC_LIB_DEVELOPMENT, "%s: unexpected circuit type: %u",
+			 __func__, circuit->circ_type);
+		exit(1);
+	}
+}
+
```

Comment:
IPv4 and IPv6 delete functions could be merge with a new `family` parameter to separate specific IPv4 from IPv6 stuff.

Reply:
Agreed. I'll merge the IPv4/IPv6 Adj-SID functions to reduce code duplication. [2]